src/twisted/web/test/test_websocket.py

```diff
@@ -0,0 +1,344 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Any, Generic, TypeVar
+from unittest import skipIf
+
+from zope.interface import implementer
+
+from twisted.internet.defer import Deferred
+from twisted.internet.error import ConnectionDone
+from twisted.internet.interfaces import IPushProducer
+from twisted.internet.protocol import Protocol, connectionDone
+from twisted.internet.testing import MemoryReactorClock
+from twisted.python.failure import Failure
+from twisted.test.iosim import ConnectionCompleter, IOPump
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.web._responses import BAD_REQUEST
+from twisted.web.client import Agent, readBody
+from twisted.web.iweb import IRequest
+from twisted.web.resource import Resource
+from twisted.web.server import NOT_DONE_YET, Request, Site
+from twisted.web.static import Data
+
+WSP = TypeVar("WSP", bound="WebSocketProtocol")
+shouldSkip = False
+
+
+class WeirdResource(Resource):
+    def render_GET(self, request: Request) -> bytes:
+        """
+        Per U{the wsproto documentation
+        <https://python-hyper.org/projects/wsproto/en/latest/api.html#wsproto.events.RejectConnection.has_body>}:
+
+            - The only scenario in which the caller receives a RejectConnection
+              with C{has_body == False} is if the peer violates sends an
+              informational status code (1xx) other than 101
+
+        This is a weird edge case so we provoke it.
+        """
+        request.setResponseCode(102)
+        return b""
+
+
+class SlowResource(Resource):
```

Comment:
Just some documentation

```suggestion
class DelayedResponseResource(Resource):
   """
   A resource that will not respond to a C{GET} request right away.
   You will need to use the C{request.write(b'RESPONSE BODY')} and C{request.finish()} to trigger a response.
   """
```